include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -0,0 +1,580 @@
+    xi -= pickRight ? leftProb : 0.0;
+
+    rcpChoiceProb = 1.0 / (pickRight ? (1.0 - leftProb) : leftProb);
+    xi *= rcpChoiceProb;
+
+    return pickRight;
+}
+
+
+// TODO: make it work in C++, ignoring problem for now
+#ifdef __HLSL_VERSION
+// @ return abs(x) if cond==true, max(x,0.0) otherwise
+template <typename T NBL_FUNC_REQUIRES(is_scalar_v<T> || is_vector_v<T>)
+T conditionalAbsOrMax(bool cond, T x, T limit);
+
+template <>
+float conditionalAbsOrMax<float>(bool cond, float x, float limit)
+{
+    const float condAbs = asfloat(asuint(x) & uint(cond ? 0x7fFFffFFu : 0xffFFffFFu));
+    return max(condAbs,limit);
+}
+
+template <>
+float32_t2 conditionalAbsOrMax<float32_t2>(bool cond, float32_t2 x, float32_t2 limit)
+{
+    const float32_t2 condAbs = asfloat(asuint(x) & select(cond, (uint32_t2)0x7fFFffFFu, (uint32_t2)0xffFFffFFu));
+    return max(condAbs,limit);
+}
+
+template <>
+float32_t3 conditionalAbsOrMax<float32_t3>(bool cond, float32_t3 x, float32_t3 limit)
+{
+    const float32_t3 condAbs = asfloat(asuint(x) & select(cond, (uint32_t3)0x7fFFffFFu, (uint32_t3)0xffFFffFFu));
+    return max(condAbs,limit);
+}
+
+template <>
+float32_t4 conditionalAbsOrMax<float32_t4>(bool cond, float32_t4 x, float32_t4 limit)
+{
+    const float32_t4 condAbs = asfloat(asuint(x) & select(cond, (uint32_t4)0x7fFFffFFu, (uint32_t4)0xffFFffFFu));
+    return max(condAbs,limit);
+}
+#endif
+
+namespace impl
+{
+struct bitFields    // need template?
+{
+    using this_t = bitFields;
+
+    static this_t create(uint32_t base, uint32_t value, uint32_t offset, uint32_t count)
+    {
+        this_t retval;
+        retval.base = base;
+        retval.value = value;
+        retval.offset = offset;
+        retval.count = count;
+        return retval;
+    }
+
+    uint32_t __insert()
+    {
+        const uint32_t shifted_masked_value = (value & ((0x1u << count) - 1u)) << offset;
+        const uint32_t lo = base & ((0x1u << offset) - 1u);
+        const uint32_t hi = base ^ lo;
+        return (hi << count) | shifted_masked_value | lo;
+    }
+
+    uint32_t __overwrite()
+    {
+#ifdef __HLSL_VERSION
+        return spirv::bitFieldInsert<uint32_t>(base, value, offset, count);
+#else
+        // TODO: double check implementation
+        const uint32_t shifted_masked_value = ~(0xffffffffu << count) << offset;
+        base &= ~shifted_masked_value;
+        return base | (value << offset);
+#endif
+    }
+
+    uint32_t base;
+    uint32_t value;
+    uint32_t offset;
+    uint32_t count;
+};
+}
+
+uint32_t bitFieldOverwrite(uint32_t base, uint32_t value, uint32_t offset, uint32_t count)
+{
+    impl::bitFields b = impl::bitFields::create(base, value, offset, count);
+    return b.__overwrite();
+}
+
+uint32_t bitFieldInsert(uint32_t base, uint32_t value, uint32_t offset, uint32_t count)
+{
+    impl::bitFields b = impl::bitFields::create(base, value, offset, count);
+    return b.__insert();
+}
```

Comment:
but we have glm and spirv for this ?

Reply:
The situation with `bitfieldInsert` is confusing to me. This mostly reflects the GLSL version we have: https://github.com/Devsh-Graphics-Programming/Nabla/blob/c7bdd2a3ec4a5a83c485533b2d68e7c9c5b8994b/include/nbl/builtin/glsl/math/functions.glsl#L263-L277

Reply:
But there's also these ones in `glsl_compat/core.hlsl`: https://github.com/Devsh-Graphics-Programming/Nabla/blob/c7bdd2a3ec4a5a83c485533b2d68e7c9c5b8994b/include/nbl/builtin/hlsl/glsl_compat/core.hlsl#L210-L226

So maybe I can remove this one? But then there's no `bitfieldOverwrite` and I don't know if they do the same thing as the GLSL version.

Reply:
@Przemog1 is responsible here, but in general:

There's should be the SPIR-V intrinsic in `spirv` namespace which provides this GLSL builtin to HLSL.

Then there should be an implementation in C++ which does the same thing, its very likely that `glm` already has one.

And finally a `nbl::hlsl::` namespace function which calls the SPIR-V or C++ depending on the language `__HLSL_VERSION`

do not remove anything, coordinate with Przemek.